src/ucp/rndv/rndv.c

```diff
@@ -784,12 +789,28 @@ ucp_rndv_init_mem_type_frag_req(ucp_worker_h worker, ucp_request_t *freq, int rn
         mem_type_ep       = worker->mem_type_ep[mem_type];
         mem_type_rma_lane = ucp_ep_config(mem_type_ep)->key.rma_bw_lanes[0];
         md_index          = ucp_ep_md_index(mem_type_ep, mem_type_rma_lane);
+        md_attr           = &mem_type_ep->worker->context->tl_mds[md_index].attr;
         ucs_assert(mem_type_rma_lane != UCP_NULL_LANE);
 
         freq->send.lane                       = mem_type_rma_lane;
         freq->send.ep                         = mem_type_ep;
         freq->send.state.dt.dt.contig.memh[0] = ucp_memh2uct(mdesc->memh, md_index);
         freq->send.state.dt.dt.contig.md_map  = UCS_BIT(md_index);
+
+        if (rkey_p != NULL && *rkey_p == NULL &&
+            md_attr->cap.flags & UCT_MD_FLAG_NEED_RKEY)
+        {
+            status = ucp_rkey_pack(mem_type_ep->worker->context, mdesc->memh,
+                                   &rkey_buffer, &rkey_size);
+            ucs_assert_always(status == UCS_OK);
+
+            status = ucp_ep_rkey_unpack(mem_type_ep, rkey_buffer, rkey_p);
+            ucs_assert_always(status == UCS_OK);
+            ucp_rkey_buffer_release(rkey_buffer);
+
+            ucs_assert_always(rkey_index != NULL);
+            memset(rkey_index, 0, UCP_MAX_LANES * sizeof(uint8_t));
+        }
```

Comment:
what is this doing?

Reply:
It's setting the `freq->send.rndv_get.rkey` to the rkey for the frag buffer so that transports that need reky (e.g. rocm_copy) can do the D2H transfer. cuda_copy didn't need rkey so previously it was set to NULL.

Reply:
can we set the rkey when fragment buffer is created and save it on mdesc?

Reply:
@yosefe I looked into this approach but not sure how feasible that is. My understanding is that the mdesc stores the list of uct mem handles for different mds and the ucp_rkey is created at runtime by calling ucp_rkey_pack and passing in the storage. We can probably have a pointer to the rkey inside ucp_mem_t struct but I'm not sure if that's a good idea because managing that memory becomes a problem. Thoughts?

Reply:
we can manage the lifetime of rkey handle same way as we do for memh itself
maybe need a set of  ucp_mpool_malloc/free functions for rndv frag pool, which will have rkey handle in addition to memh

Reply:
we'd like to avoid rkey pack/unpack for every fragment, if possible, for performance reason

Reply:
Is it okay to push this improvement in a follow up PR? a) This codepath doesn't get triggered for non-ROCm memory and b) the performance is bottlenecked by rocm copy performance which is much larger than cost of packing rkey.

Reply:
@yosefe I have rebased onto latest master and squashed the changes.